### DISCUSSION TITLE
issue in using SAHI with Detectron2

### DISCUSSION THREAD

**Discussion by imadgohar** [obss/sahi#611] [Q&A] [answered]
@fcakyon 
I Installed detectron2 on windows server and then installed sahi. After importing the detectron2 and sahi I am unable to export cfg 
ModuleNotFoundError: No module named 'sahi.utils.detectron2' although everything is installed and I can import it. what I am missing here. 

I am following the tutorial on 
medium https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f

I am on windows server with the following settings
pytorch 1.9
cuda 11.1
python 3.9 

Thanks.

**Comment by fcakyon** [comment]
What is your sahi version?

**Comment by fcakyon** [comment]
Are you sure you are running the code in the correct environment (in the environment that sahi is installed)?

**Comment by imadgohar** [comment]
Thanks for your response,

I just installed with this command "conda install sahi -c conda-forge". Yes I am in the same environment I can import sahi, detectron2 separately. Also I tried sahi.utils it is working but sahi.utils.detectron2 in not installed I don't know why.

**Comment by fcakyon** [comment]
Can you tell us the sahi version by `conda list sahi`?

**Comment by fcakyon** [comment]
Have you tried:

```python
import sahi.utils
from sahi.utils.detectron2 import export_cfg_as_yaml 
```

**Comment by imadgohar** [comment]
I tried

import sahi and then 
from sahi.utils.detectron2 import export_cfg_as_yaml 
export_cfg_as_yaml(cfg, export_path='config.yaml')

I will try the above and will let you know

**Comment by imadgohar** [comment]
Sahi version is 0.10.6


> Can you tell us the sahi version by `conda list sahi`?

**Comment by fcakyon** [comment]
I am not having any error when running these lines:

```python
import sahi
from sahi.utils.detectron2 import export_cfg_as_yaml
```

**Comment by imadgohar** [comment]
> I am not having any error when running these lines:
> 
> ```python
> import sahi
> from sahi.utils.detectron2 import export_cfg_as_yaml
> ```

The issue solved with installing new environment and install sahi with pip instead of conda "pip install -U sahi".

import sahi.utils.detectron2 is importing now but again i am getting this error. 
![a](https://user-images.githubusercontent.com/45845910/192505046-8bfe50bb-6f31-4067-a7ea-de16999fcd0a.PNG)

**Comment by fcakyon** [comment]
@kadirnar can you test it in your free time? :sweat_smile:

**Comment by kadirnar** [comment]
@imadgohar 
Can you tell me the steps of installing detectron?

**Comment by fcakyon** [comment]
> @imadgohar Can you tell me the steps of installing detectron?

@kadirnar you can follow the steps from blogpost: https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f

**Comment by imadgohar** [comment]
> @imadgohar Can you tell me the steps of installing detectron?

I am on windows server, the steps to install detectron is
1. Install build tools
2. Install cython
3. Pycocotools
4. Then detectron2
I already trained my model on that I want to use sahi for inference to check if it can help in detecting small objects (in high Res images).

After that I installed sahi with conda but sahi.utils.detectron2 was not importing the. I created a new env installed detectron2 again and then install sahi with pip U command now I can import it but again when I run the predict command detectron2 is still not found.

**Comment by imadgohar** [comment]
> > @imadgohar Can you tell me the steps of installing detectron?
> 
> @kadirnar you can follow the steps from blogpost: https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f

I started following this, but I think the command here to install detectron2 is for Linux..yes?
On windows it is not that much simple to install.

**Comment by fcakyon** [comment]
> > > @imadgohar Can you tell me the steps of installing detectron?
> > 
> > 
> > @kadirnar you can follow the steps from blogpost: https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f
> 
> I started following this, but I think the command here to install detectron2 is for Linux..yes? On windows it is not that much simple to install.

Yes, currently we don't support detectron2 for windows, our tests run for linux only.

**Comment by imadgohar** [comment]
> > > > @imadgohar Can you tell me the steps of installing detectron?
> > > 
> > > 
> > > @kadirnar you can follow the steps from blogpost: https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f
> > 
> > 
> > I started following this, but I think the command here to install detectron2 is for Linux..yes? On windows it is not that much simple to install.
> 
> Yes, currently we don't support detectron2 for windows, our tests run for linux only.

Hmmm...means I can't use it on windows even detectron2 is installed and working.

But I am wonder I did run it with yolov5 on windows server (Although the results are starnge) but at least I am getting it.

**Comment by fcakyon** [comment]
> Hmmm...means I can't use it on windows even detectron2 is installed and working.

No it means we are not testing detectron2+sahi support. If you manage to install it on windows it should work fine :+1:

**Comment by kadirnar** [comment]
If you create a repo on this issue, I can help. @imadgohar

**Comment by aymanaboghonim** [comment]
It is a detectron2 related issue and I solved it by :
1. Creating a virtual env with python 3.6 and activate it
2. Installing appropriate Torch and Cuda toolkit
3. Detectron2 dependencies like   Pycocotools and OpenCv
4. Install detectron2 and Sahi
lets try and see.

**Reply by imadgohar** [reply]
Exactly same thing I am doing just python version is different but I don't know it's a problem.

**Comment by imadgohar** [comment]
Oh great to hear that,

Can you please tell me which version of cuda and torch you installed?

Second thing, after installing the detectron2 just install sahi right?

Thanks.

**Reply by aymanaboghonim** [reply]
I am using torch 1.9 and cuda 10.2 
`conda install pytorch==1.9.0 torchvision==0.10.0 torchaudio==0.9.0 cudatoolkit=10.2 -c pytorch`.
yes, installing sahi after installing detectron2 .

**Reply by imadgohar** [reply]
let me check with this env if it works. thanks for the update.

**Comment by imadgohar** [comment]
> If you create a repo on this issue, I can help. @imadgohar

Do I need to create another repo?

**Reply by kadirnar** [reply]
Are you asking for windows system?

**Reply by imadgohar** [reply]
> Are you asking for windows system?

Yes.

**Reply by kadirnar** [reply]
Can you create a new repo by looking at this repo?
https://github.com/kadirnar/Yolov6-SAHI

**Comment by fcakyon** [comment]
How did you solve this issue? @imadgohar

**Reply by imadgohar** [reply]
Actually the issue was with detectron2 installation. I switched from windows to colab and then install these versions of cuda and torch.
!pip install torch==1.10.1+cu111 torchvision==0.11.2+cu111 -f https://download.pytorch.org/whl/torch_stable.html
!pip install -q pyyaml==5.3.1
!pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cu111/torch1.10/index.html
!pip install -U sahi

this setup works fine.

**Reply by fcakyon** [reply]
thanks a lot for the details!

**Comment by fcakyon** [comment, marked as answer]
Actually the issue was with detectron2 installation. I switched from windows to colab and then install these versions of cuda and torch.
!pip install torch==1.10.1+cu111 torchvision==0.11.2+cu111 -f https://download.pytorch.org/whl/torch_stable.html
!pip install -q pyyaml==5.3.1
!pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cu111/torch1.10/index.html
!pip install -U sahi

this setup works fine.

solved by @imadgohar

**Reply by AbdullahJirjees** [reply]
I am getting error when I use 
from sahi.model import DetectionModel 
the error is No module named sahi.model 
Any idea?

**Reply by fcakyon** [reply]
try `from sahi.models.base import DetectionModel`

**Reply by AbdullahJirjees** [reply]
Worked! thanks a lot

**Reply by fcakyon** [reply]
I am glad to hear that!

**Reply by AbdullahJirjees** [reply]
Hello @fcakyon 
I am getting error with the dataloader, i am using icevision on google colab, any idea what is going on here? 

![Screenshot 2022-11-06 205628](https://user-images.githubusercontent.com/13992604/200223547-88bdf080-4508-4007-91b1-64dca8ad2b56.png)

**Reply by fcakyon** [reply]
This seems like a icevision-specific error. You should open an issue on that repo :)

**Comment by imadgohar** [comment]
@fcakyon 
I have dataset with image size 5k * 3k, I divided the images into different scales let's say 800*800 and 640*640 and trained the model. 

During inference with sahi, I am using the full resolution images.

When I pass image_size 5k to the sahi model during inference the mAP is increased by about 6 to 10% but the time it takes to process one image is 10x higher than without passing image_size.

What is your suggestions on this? Is it ok or I don't need to pass this parameter?

**Reply by imadgohar** [reply]
@fcakyon In short I mean that when I don't pass the image_size parameter it will just scale the image as mentioned in the code.
 image_size: int
            Input image size for each inference (image is scaled by preserving asp. rat.).
In my case inference time increasing but results improving.

**Reply by fcakyon** [reply]
It is expected. As the number of slices increases, accuracy and inference time also increase. 

If you want inference time to get shorter, you can: 
- increase the slice_size, 
- decrease the overlap_ratio, 
- increase the model_confidence_threshold, 
- decrease the iamge_size, 
- use NMS postprocess

**Reply by imadgohar** [reply]
Thanks @fcakyon for your suggestions, Yes I tried these options.

**Reply by fcakyon** [reply]
We are working on more efficient inference pipelines (with batching and threaded queueing) on the `v0.12` branch, but it is still experimental, includes breaking changes, and is constantly updated. You can wait for a stable release or implement your custom inference pipelines.

**Reply by imadgohar** [reply]
Nice to hear that.

**Reply by imadgohar** [reply]
@fcakyon Is there any changes to the code. because I run the code successfully around two weeks ago with detectron2 although the output was all 0 but code was running. I thought that maybe their is an issue with my model. So I trained by model again and now I am running the same code but it giving me Assertion Error. You can see in the screenshot.
![error](https://user-images.githubusercontent.com/45845910/204703515-88e6b286-2326-4266-998f-a6ad08fcd20c.PNG)

Also, from the same setup this was the output, atleast it was running.

![last-model](https://user-images.githubusercontent.com/45845910/204703643-d56a7580-5a7d-4d3b-81c5-c72f6cf6c843.PNG)

Can you provide an updated running notebook with detecctron2. Because I always got trouble using SAHI with detectron2 with yolov5 it is working fine.

**Reply by fcakyon** [reply]
Can you post the full error traceback so that we can find the line causing this error?

I can release a fix today, once I know the related line :)

**Reply by imadgohar** [reply]
Here it is, 
![error2](https://user-images.githubusercontent.com/45845910/204716360-62086770-0cad-4491-b4ea-40163cd7e55c.PNG)

**Reply by fcakyon** [reply]
I have found the issue, will release a fix at most tomorrow 👍

**Reply by imadgohar** [reply]
Thanks @fcakyon will wait for it.

**Reply by fcakyon** [reply]
This pr is fixing this issue: https://github.com/obss/sahi/pull/763

Will release `sahi==0.11.6` once the tests pass 👍

**Reply by imadgohar** [reply]
Thanks for the update @fcakyon

**Reply by fcakyon** [reply]
@imadgohar, thanks a lot for bringing up this issue!

It is now fixed in the latest version: https://github.com/obss/sahi/releases/tag/0.11.6

**Reply by imadgohar** [reply]
Thanks @fcakyon the issue is solved now but my results are still 0 and very poor detections in visualization, What should be the possible reason I am just wonder bcs on validation set I am getting good results without sahi. The model trained on yolov5 with same setup is performing well with sahi but not the case with detectron2.
![er](https://user-images.githubusercontent.com/45845910/205232636-9da33cc8-3a1e-48b6-a59c-364b618ff236.PNG)

**Reply by fcakyon** [reply]
Most probably your category ids are not correctly setup, they should start from 0 instead of 1.

**Reply by imadgohar** [reply]
For re-mapping my id's I am using your code:

from sahi.utils.coco import Coco
from sahi.utils.file import save_json

coco = Coco.from_coco_dict_or_path("test-updated.json")

desired_name2id = {
  "label1": 0,
  "Label2": 1,
  "Label3": 2,
  "label4": 3,
  "Label5": 4
}
coco.update_categories(desired_name2id)

save_json(coco.json, "test-updated-v0-4.json")

But still everything is 0.

**Reply by imadgohar** [reply]
Just to compare the category Id's, from my file it is fro 0 to 4 
![1](https://user-images.githubusercontent.com/45845910/205238690-c7e8af08-da56-4754-a2e4-381bb68bffbc.PNG)
 and from the generated result file is very strange not seen 0 or 1 here 21,27,68 etc.
![2](https://user-images.githubusercontent.com/45845910/205239021-4a082bfc-9a6e-4e82-bf24-1973b771b51b.PNG)

**Reply by imadgohar** [reply]
From this I get the cfg
from detectron2.config import get_cfg
cfg = get_cfg()
cfg.MODEL.RETINANET.NUM_CLASSES = 5 
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.01

from sahi.utils.detectron2 import export_cfg_as_yaml 
export_cfg_as_yaml(cfg, export_path='config.yaml')

**Reply by imadgohar** [reply]
before starting predictions I am receiving this warning, One thing that come in my mind is I think during training my id's map from 1-5 but I think it should be no more a problem if I re-map the id's beore using sahi to 0-4.
WARNING:sahi.models.detectron2:tuple index out of range

**Reply by imadgohar** [reply]
@fcakyon @kadirnar 
I run your example sahi with detectron2 colab notebook, It is working fine. 

1. Just to ask you used the model from here 
model_path = Detectron2TestConstants.FASTERCNN_MODEL_ZOO_NAME
but I trained fasterrcnn with cutom data and i am passing my model like this
model_path = "model_final.pth"

2. In config file you are passing the same variable while I am using the extracted config file using this 
3. from detectron2.config import get_cfg
cfg = get_cfg()
cfg.MODEL.RETINANET.NUM_CLASSES = 5
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.01

from sahi.utils.detectron2 import export_cfg_as_yaml
export_cfg_as_yaml(cfg, export_path='config.yaml')

Then why my prediction results are all zero? mapping is also correct. it is starting from 0.

**Reply by fcakyon** [reply]
@imadgohar what is your detectron2 model type? Can you paste the link of your base config from the detectron2 repo?

**Reply by imadgohar** [reply]
Yes sure
For FasterRCNN it is https://dl.fbaipublicfiles.com/detectron2/COCO-Detection/faster_rcnn_R_50_FPN_3x/137849458/model_final_280758.pkl 
and for RetinaNet it is https://dl.fbaipublicfiles.com/detectron2/COCO-Detection/retinanet_R_50_FPN_3x/190397829/model_final_5bd44e.pkl
I am using for RetinaNet config:
#cfg.MODEL.RETINANET.NUM_CLASSES = 5 
#cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5
and for FasterRCNN:
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 5
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.5

**Reply by imadgohar** [reply]
Thanks @fcakyon for all your responses. I figured out the problem it's the config file, I am not getting zero result. Still I am improving it but the problem is at least figured out. We need to change many parameters in the config file.

**Reply by fcakyon** [reply]
@imadgohar can you explain your solution? Which parameters you had to change? Maybe we can update the medium post accordingly :+1:

**Reply by imadgohar** [reply]
> @imadgohar can you explain your solution? Which parameters you had to change? Maybe we can update the medium post accordingly 👍

Sorry for being late, In order to run detectron2 with custom dataset and SAHI you need to create config file and update all the parameters used during for custom training.

The provided code outputs the config file but it is not enough for correct results.
